### ISSUE TITLE
Synchronized playback over a network

### ISSUE THREAD

**Issue by jdb-vsn** [mpv-player/mpv#1272]
Description source: http://www.mplayerhq.hu/DOCS/HTML/en/networksync.html

> Multiple instances of MPlayer can synchronize playback over a network. This is useful for creating "video walls" with multiple screens controlled by different computers. Each MPlayer instance can play a different video, but they all will try to stay at the same time offset in the file. It is recommended but not necessary to encode the video files using the same codec and parameters.
> 
> The relevant options are -udp-master, -udp-slave, -udp-ip, -udp-port, and -udp-seek-threshold.
> 
> If -udp-master is given, MPlayer sends a datagram to -udp-ip (default: 127.0.0.1) on -udp-port (default: 23867) just before playing each frame. The datagram indicates the master's position in the file. If -udp-slave is given, MPlayer listens on -udp-ip/-udp-port and matches the master's position. Setting -udp-ip to the master's broadcast address allows multiple slaves having the same broadcast address to sync to the master. Note that this feature assumes an ethernet-like low-latency network connection. Your mileage may vary on high latency networks.

Mentioned on issue #898.  

Also on https://github.com/mpv-player/mpv/issues/252#issuecomment-24864050.

**Comment by artwait** [comment]
I'm using the udp sync feature in mplayer right now for a video wall, but I'm running multiple slaves on a single machine (rather than across a network). The challenge with this is that the UDP port is blocked by the first slave, thus subsequent slaves are not able to listen. I've got a hacky work-around to this, but it would be nice to also have a solution that uses a non-network approach for sending timestamps between instances on a single machine. Just a thought.

**Comment by wm4** [comment]
> it would be nice to also have a solution that uses a non-network approach for sending timestamps between instances on a single machine.

Do you have a specific mechanism in mind?

**Comment by lionep** [comment]
:+1:

**Comment by mgcrea** [comment]
:+1:

**Comment by gcoupelant** [comment]
:+1: This would really be awesome.

**Comment by smarek** [comment]
Hey guys, any progress on this feature please?

**Comment by wm4** [comment]
Nobody is working on it. It could now be implemented without mpv changes via one of the following mechanisms:
- as Lua script
- as a cplugin
- as a wrapper application using libmpv

**Comment by smarek** [comment]
@wm4 cool, so any network synchronized protocol and wrapper around mpv should suffice? 
Maybe using `--input-ipc-server` and any daemon that converts network issued command into mpv json command (or just dumb pass-through) could work.

Also I've found this, but theres no documentation. https://github.com/derlaft/synco
Anybody tried to use that?

**Comment by rossy** [comment]
@smarek You could also look at http://syncplay.pl/

I don't know if it's accurate enough to use as a video wall (like MPlayer's UDP-sync apparently was,) but it definitely can sync multiple mpv instances over a network.

**Comment by Collocalini** [comment]
I've recently did just what Smarek offered - used --input-ipc-server mode to synchronize mpv and osmcplayer from kodi on raspberry pi. Supports play/pause detection. I used it mostly to show video from raspberry pi but take audio from a better audio card on a separate computer. Not perfect but I find it watchable most of the time.
[mpvsync.zip](https://github.com/mpv-player/mpv/files/975628/mpvsync.zip)

**Comment by bitbyt3r** [comment]
I've been trying to do this through libmpv and have had some success, however it isn't quite frame perfect. I'm using a loop similar to this:
`

     mpv_observe_property(player0, 0, "time-pos", MPV_FORMAT_DOUBLE);

     while (1) {
        mpv_event *event = mpv_wait_event(ctx[0], 1000);

        if (event->event_id == MPV_EVENT_PROPERTY_CHANGE) {
            mpv_event_property *prop = (mpv_event_property *)event->data;
            if (strcmp(prop->name, "time-pos") == 0) {
                double pos = *(double*)prop->data;
                for (int i=1; i<instances; i++) {
                    mpv_set_property(player1, "time-pos", MPV_FORMAT_DOUBLE, &pos);
                }
            }
        }

        if (event->event_id == MPV_EVENT_SHUTDOWN)
            break;
    }
`

Basically I subscribe to changes of the "time-pos" property, then set the time-pos of every slave player instance. This gets really close to what I want, but sometimes doubles a frame or skips a single frame and I'll end up off by one for a bit before returning to the correct timing. Is there a better method that I could use here? The time-pos property appears to update once per frame, so it is a bit course for this.

**Comment by wm4** [comment]
A  better method would be adjusting playback speed for small deviations. Also "time-pos" is somewhat asynchronously reported, so it can definitely happen that it's off by a frame or more.

**Comment by hansinator** [comment]
Hi,
I need network sync support like in mplayer but with the ability sync the playlist position, i.e. the slaves should follow the master if the master plays the another video in the playlist. I am currently using mplayer, but I would need to modify the mplayer network sync anyway (and make it incompatible to the upstream version), so I've decided it'd be a good opportunity to implement this whole thing in mpv. I am not convinced by a separate executable, plugin or script. I'd like to add support directly into mpv so that the feature is readily available via command line and for library users.

Do you have any suggestions or pointers into the right direction before I start? I have read that the blocking socket read behaviour on the slave side is undesired. Does anyone have an idea what a better solution might look like? However, I believe this is an implementation detail that can be changed at any time without affecting the network packet format or user experience.

**Comment by nixternal** [comment]
I built mpv with the network sync stuff, but I am clueless on how to use it properly. Any help there will be greatly appreciated. Also, anyone done any further work since last year? mplayer is great and all, but when utilizing Intel graphics & libvdpau-va-gl anything other than h264 is a no go.

**Comment by Collocalini** [comment]
2nixternal:
use 
{ "command": ["get_property", "speed"], "request_id": "asc" } to get playback speed
{ "command": ["get_property", "time-pos"], "request_id": "gtp" } to get current playback position

The request_id could be arbitrary.

{ "command": ["set_property", "speed", 1.0]} to set speed
{ "command": ["set_property", "time-pos", 0.0]} to set playback position

Commands should be sent through a socket. Run mpv like this:
mpv --input-ipc-server=/tmp/mpvsocket file.mp4

for more see section JSON IPC from https://mpv.io/manual/master/

**Comment by EternityForest** [comment]
Regular NTP can be pretty darn good on a wired network.  A really standard, interoperable way could be to just use the system time.  It's probably never going to be worse than your ping time, and WiFi almost always let's a few 2ms or so packets through.  Daemons like Chrony have this stuff all figured out, why reinvent?  Even Windows can apparently do 1ms sync?

Using a --wallclock-start=TIMESTAMP,  you could tell MPV that you want to play synced to wallclock, with the start of the video at exactly that time.  If the command needs time to process and it can't hit the time you specified, or it's in the future, you skip or delay as needed so that the current position is always what it would be if you started exact then.

With 1 second of video locked to 1 second of real time, everything is perfect as long as they get the same time.  You can even start them at different times.  No communication is actually needed here, you could just have GPS.

Network issues will not cause stuttering as long as they are brief enough that the clocks stat synced.

This would be troublesome manually, but from there it's easy to add a tiny bit of automation with 2 really simple packets.   The first one gets a --sync-group=name argument, and the second one gets a --sync-join argument.

The second sends a multicast packet asking for the group master's start time, putting the name in the UDP.  Once it gets that, no communication is needed.

**Comment by pyfave** [comment]
There is a big need for this and osc frame search.

**Comment by nixternal** [comment]
Yeah, I have machines in the field that have been running for years with mplayer -udp-master & -udp-slave, all running sound. Only time we have ever noticed any sort of drift is poorly made videos. When we keyint=1 so every frame is keyed, the only way you would detect any drift is with a picture or record a video & go frame-by-frame.

**Comment by guillaumekh** [comment]
@nixternal that's interesting. Do you do anything to control audio backend latency — i.e. anything that would ensure that the audio backend on each machine running mpv is within the same range? Come to think of it, using identical machines running same OS would probably qualify.

@Frenzie i wasn't really commenting on UDP latency. My point was only aimed at the latency that audio backends introduce — e.g.g PulseAudio can easily introduce >100ms latency since low latency was never one of its design goals.

**Comment by nixternal** [comment]
@guillaumekh we do nothing at all. Our machines are built upon the Intel NUC so there is a chance they are all the same, but usually just similar on the rental side. All running some LTS version of Ubuntu. mPlayer isn't bad, but with Intel you can't do h.265 or anything bigger than 4k. The mplayer-vaapi project is long gone & dead, and the whole vdpau-vaapi stuff is a mess. We are now using Pulse as our backend, but used Alsa prior to using 18.04 iirc.

**Comment by nixternal** [comment]
Yeah, we looked at all of the Dante stuff as well, and it still confuses me about as much as DMX does, but they UDP sync with mplayer is so good, when you keyint=1 every frame. If you don't, yeah there "could" be a little noticeable drift.

`ffmpeg -i input.mp4 -c:v libx264 -x264opts keyint=1 -crf 15 -c:a copy output.mp4`

We run this quick little script on every video & they have been flawless. Like I tell my clients, garbage in, garbage out which has been the case.

**Comment by chidea** [comment]
I've been working around for years to implement this with my main language, Rust.
Since my clients are using unstable wireless networks, what I stumbled upon was relying on very short UDP packets instead of long packets with timestamps.
I am currently planning to implement its bare minimums into a lua plugin scripts in this year.

**Comment by mr-berndt** [comment]
I just recently had to synchronize a number of talking heads and did so with a lua script and Claude code's help. I did not optimize it to the very end but it keeps quiet a useable sync so far. You might want to check it out. https://github.com/mr-berndt/mpv-udp-framesync
